### DISCUSSION TITLE
From Numpy Array to  TC Tensor and vice versa

### DISCUSSION THREAD

**Discussion by MuhamedEssam** [haydnv/tinychain#103] [General]
if i have  `X` with type `np.array` how can i convert it to be Tinychain Tensor format 
So i can use Tc.tensor methods ? 

and for the opposite should i use `.flatten.tolist()`  then cast it by  `np.array()`

**Comment by MuhamedEssam** [comment]
Also  do we have  `.shape` Function over here in tinychain ?

**Comment by MuhamedEssam** [comment]
i tested this in numpy and it worked  
```
        self.dbeta = 2/n_samples*(np.einsum("ij,ki->kj", self.beta,
            np.einsum("ij,ki->kj", self.X, self.X.T)) 
                                  - np.einsum("ij,ki->kj", self.y, self.X.T))  

```
can we say that 
 ```
       d_weights = 2/len(inputs)*(tc.tensor.einsum("ij,ki->kj", weights,
                                                    tc.tensor.einsum("ij,ki->kj", inputs, inputs.transpose))-
                                                       tc.tensor.einsum("ij,ki->kj", labels, inputs.transpose)) 
```

**Comment by haydnv** [comment]
Hi Muhamed, there is a `Tensor.shape` function which returns a `Tuple` of `U64`s and I just committed a change (in #102) so that `Tuple` now supports `len` and `iter` when its contents are known at compile-time. So, the only issues I can possibly see with your code are `len(inputs)` (it might be better to call `inputs.shape[0]`, for a `Tensor`, or `inputs.len()`, for a `Tuple` which may not be resolved at compile-time) and `inputs.transpose()` (note: `transpose` is a method and needs the parentheses to call it).

In general you can load a `Tensor` from an `nparray` `x` by calling `Tensor.load(x.shape, tc.F32, x.flatten().tolist())` where you should replace `F32` with whatever data type you're expecting. Eventually the Python client will be smart enough to support NumPy and Pandas types (almost) automatically but this is still a few months away.

**Reply by MuhamedEssam** [reply]
but  when we flatten then we will lose information about width and hight and it will be replaced by size information only ?  you got me ?

**Reply by haydnv** [reply]
The purpose of providing `x.shape` is to set the shape of the `Tensor`, so you won't lose that information. The reason you need to call `.flatten().tolist()` is that the TinyChain client doesn't know about NumPy (so that client users won't have to install NumPy in order to use TinyChain) and it encodes, stores, and transmits tensor elements as one long sequence for reasons of efficiency. You can check that the shape is treated correctly by using any of the `Tensor` methods that accept an `axis` argument, e.g. `d_weights.reduce_sum(axis=-1)`.
